spacy/cli/project/parallel.py

```diff
@@ -0,0 +1,442 @@
+        )
+        first = True
+        while any(
+            cmd_info.state.name in ("pending", "starting", "running")
+            for cmd_info in cmd_infos
+        ):
+            try:
+                mess: Dict[str, Union[str, int]] = parallel_group_status_queue.get(
+                    timeout=PARALLEL_GROUP_STATUS_INTERVAL * 20
+                )
+            except Exception:
+                # No more messages are being received: the whole group has hung
+                for other_cmd_info in (
+                    c for c in cmd_infos if c.state.name in ("starting", "running")
+                ):
+                    other_cmd_info.change_state("hung")
+                for other_cmd_info in (
+                    c for c in cmd_infos if c.state.name == "pending"
+                ):
+                    other_cmd_info.change_state("cancelled")
+                break
+            cmd_info = cmd_infos[cast(int, mess["cmd_index"])]
+            if mess["type"] in ("started", "keepalive"):
+                cmd_info.last_keepalive_time = int(time())
+            for other_cmd_info in (
+                c for c in cmd_infos if c.state.name in ("starting", "running")
+            ):
+                if (
+                    other_cmd_info.last_keepalive_time is not None
+                    and time() - other_cmd_info.last_keepalive_time
+                    > PARALLEL_GROUP_STATUS_INTERVAL * 20
+                ):
+                    # a specific command has hung
+                    other_cmd_info.change_state("hung")
+            if mess["type"] == "started":
+                cmd_info.change_state("running")
+                cmd_info.running_os_cmd_index = cast(int, mess["os_cmd_index"])
+                cmd_info.pid = cast(int, mess["pid"])
+            if mess["type"] == "completed":
+                cmd_info.rc = cast(int, mess["rc"])
+                if cmd_info.rc == 0:
+                    cmd_info.change_state("succeeded")
+                    if not dry:
+                        update_lockfile(current_dir, cmd_info.cmd)
+                    next_worker_process = next(worker_process_iterator, None)
+                    if next_worker_process is not None:
+                        _start_worker_process(next_worker_process, proc_to_cmd_infos)
+                elif cmd_info.rc > 0:
+                    cmd_info.change_state("failed")
+                else:
+                    cmd_info.change_state("terminated")
+                cmd_info.console_output = cast(str, mess["console_output"])
+            if any(
+                c for c in cmd_infos if c.state.name in ("failed", "terminated", "hung")
+            ):
+                # a command in the group hasn't succeeded, so terminate/cancel the rest
+                for other_cmd_info in (
+                    c for c in cmd_infos if c.state.name == "running"
+                ):
+                    try:
+                        os.kill(cast(int, other_cmd_info.pid), SIGTERM)
```

Comment:
Why are you `cast`ing here?

Reply:
The type of the variable is `Optional[int]`.

Reply:
Ah, `mypy` complains indeed otherwise.